File: tests/support/harness.py
"""Full-stack test harness: MCP server + broker + live loopback UI.

Everything that crosses a boundary is recorded into a :class:`LeakScanner`, so
any test can end with ``harness.scanner.assert_clean(canary)``.
"""

from __future__ import annotations

import http.client
import io
import json
import urllib.parse
from dataclasses import dataclass, field
from typing import Any

from tests.support.canary import LeakScanner
from veil.adapters.registry import AdapterRegistry
from veil.broker import SecretBroker
from veil.config import VeilConfig
from veil.logging_ import AuditLogger
from veil.mcp_server.server import MCPServer
from veil.ui.server import SecureInputUI


@dataclass
class HttpResponse:
    status: int
    headers: dict[str, str]
    body: str


@dataclass
class Harness:
    config: VeilConfig
    registry: AdapterRegistry
    broker: SecretBroker
    server: MCPServer
    ui: SecureInputUI
    log_stream: io.StringIO
    log_records: list[dict[str, Any]]
    scanner: LeakScanner = field(default_factory=LeakScanner)
    _next_id: int = 0

    # -- MCP ---------------------------------------------------------------

    def rpc(self, method: str, params: dict[str, Any] | None = None) -> dict[str, Any]:
        self._next_id += 1
        message = {"jsonrpc": "2.0", "id": self._next_id, "method": method}
        if params is not None:
            message["params"] = params
        self.scanner.add("mcp_traffic", json.dumps(message))
        self.scanner.add("mcp_tool_arguments", json.dumps(params or {}))
        response = self.server.handle_message(message)
        rendered = json.dumps(response)
        self.scanner.add("mcp_traffic", rendered)
        self.scanner.add("mcp_tool_results", rendered)
        assert response is not None
        return response

    def call_tool(self, name: str, arguments: dict[str, Any]) -> dict[str, Any]:
        response = self.rpc("tools/call", {"name": name, "arguments": arguments})
        result = response.get("result")
        assert result is not None, response
        payload: dict[str, Any] = result["structuredContent"]
        return payload

    # -- HTTP --------------------------------------------------------------

    def _connect(self) -> http.client.HTTPConnection:
        assert self.ui.base_url
        parsed = urllib.parse.urlsplit(self.ui.base_url)
        assert parsed.hostname and parsed.port
        return http.client.HTTPConnection(parsed.hostname, parsed.port, timeout=10)

    def get(self, path: str) -> HttpResponse:
        self.scanner.add("http_urls", path)
        connection = self._connect()
        try:
            connection.request("GET", path, headers={"Host": "127.0.0.1"})
            raw = connection.getresponse()
            body = raw.read().decode("utf-8", "replace")
            response = HttpResponse(raw.status, dict(raw.getheaders()), body)
        finally:
            connection.close()
        self.scanner.add("ui_html", response.body)
        return response

    def post(self, path: str, fields: dict[str, str]) -> HttpResponse:
        self.scanner.add("http_urls", path)
        body = urllib.parse.urlencode(fields)
        connection = self._connect()
        try:
            connection.request(
                "POST",
                path,
                body=body.encode("utf-8"),
                headers={
                    "Host": "127.0.0.1",
                    "Content-Type": "application/x-www-form-urlencoded",
                    "Content-Length": str(len(body.encode("utf-8"))),
                },
            )
            raw = connection.getresponse()
            payload = raw.read().decode("utf-8", "replace")
            response = HttpResponse(raw.status, dict(raw.getheaders()), payload)
        finally:
            connection.close()
        if response.status != 303:
            self.scanner.add("ui_html", response.body)
        return response

    # -- flows -------------------------------------------------------------

    def path_for(self, payload: dict[str, Any]) -> str:
        url = payload["authorization_url"]
        return urllib.parse.urlsplit(url).path

    def submit_secret(self, payload: dict[str, Any], value: str) -> HttpResponse:
        return self.post(f"{self.path_for(payload)}/submit", {"secret": value})

    def confirm(self, payload: dict[str, Any]) -> HttpResponse:
        path = self.path_for(payload)
        page = self.get(path)
        token = _hidden_field(page.body, "confirm_token")
        assert token, "stage B page did not render a confirmation token"
        return self.post(f"{path}/confirm", {"confirm_token": token})

    def cancel(self, payload: dict[str, Any]) -> HttpResponse:
        return self.post(f"{self.path_for(payload)}/cancel", {})

    def status(self, request_id: str) -> dict[str, Any]:
        return self.call_tool("secret.status", {"request_id": request_id})

    def finish_flow(self, payload: dict[str, Any], value: str) -> dict[str, Any]:
        """Stage A (+ Stage B when required) through to a terminal state."""

        self.get(self.path_for(payload))
        self.submit_secret(payload, value)
        status = self.status(payload["request_id"])
        if status["state"] == "AWAITING_EXECUTION_CONFIRMATION":
            self.confirm(payload)
            status = self.status(payload["request_id"])
        return status

    def collect_logs(self) -> None:
        self.scanner.add("application_logs", self.log_stream.getvalue())
        self.scanner.add("audit_records", json.dumps(self.log_records))


def build_harness(
    config: VeilConfig,
    registry: AdapterRegistry,
    *,
    start_ui: bool = True,
) -> Harness:
    log_stream = io.StringIO()
    records: list[dict[str, Any]] = []
    logger = AuditLogger(stream=log_stream, sink=records.append)
    broker = SecretBroker(config, registry, logger=logger)
    ui = SecureInputUI(broker, config, logger=logger)
    server = MCPServer(broker, registry, logger=logger)
    if start_ui:
        ui.start()
    return Harness(
        config=config,
        registry=registry,
        broker=broker,
        server=server,
        ui=ui,
        log_stream=log_stream,
        log_records=records,
    )


def _hidden_field(html: str, name: str) -> str | None:
    marker = f'name="{name}" value="'
    index = html.find(marker)
    if index < 0:
        return None
    start = index + len(marker)
    end = html.find('"', start)
    return html[start:end] if end > start else None

File: test/support/harness.ts
/**
 * Full-stack test harness: MCP server + broker + live loopback UI.
 *
 * Everything that crosses a boundary is recorded into a `LeakScanner`, so any
 * test can end with `harness.scanner.assertClean(canary)`.
 */

import { AdapterRegistry } from '../../src/adapters/registry.js';
import { SecretBroker } from '../../src/broker.js';
import { defaultConfig, type VeilConfig } from '../../src/config.js';
import { AuditLogger } from '../../src/logging.js';
import { MCPServer, type JsonRpcResponse } from '../../src/mcp/server.js';
import { SecureInputUI } from '../../src/ui/server.js';
import { LeakScanner } from './canary.js';

export interface HttpResult {
  readonly status: number;
  readonly headers: Headers;
  readonly body: string;
}

/**
 * Test configuration.
 *
 * The harness plays the human, so it needs the link a browser would receive;
 * `test/security/authorization/out-of-band.test.ts` covers the production
 * default, where the agent never sees it.
 */
export function testConfig(overrides: Partial<VeilConfig> = {}): VeilConfig {
  return defaultConfig({
    requestTtlSeconds: 60,
    adapterTimeoutSeconds: 5,
    uiHost: '127.0.0.1',
    uiPort: 0,
    openBrowser: false,
    discloseAuthorizationUrl: true,
    ...overrides,
  });
}

export class Harness {
  readonly scanner = new LeakScanner();
  readonly logRecords: Record<string, unknown>[] = [];
  logText = '';
  readonly broker: SecretBroker;
  readonly ui: SecureInputUI;
  readonly server: MCPServer;
  #nextId = 0;

  private constructor(
    readonly config: VeilConfig,
    readonly registry: AdapterRegistry,
  ) {
    const logger = new AuditLogger({
      stream: {
        write: (chunk: string): boolean => {
          this.logText += chunk;
          return true;
        },
      },
      sink: (record) => this.logRecords.push(record),
    });
    this.broker = new SecretBroker(config, registry, { logger });
    this.ui = new SecureInputUI(this.broker, config, { logger });
    this.server = new MCPServer(this.broker, registry, { logger });
  }

  static async start(config: VeilConfig, registry: AdapterRegistry): Promise<Harness> {
    const harness = new Harness(config, registry);
    await harness.ui.start();
    return harness;
  }

  async stop(): Promise<void> {
    this.broker.shutdown();
    await this.ui.stop();
  }

  // -- MCP --------------------------------------------------------------------

  async rpc(method: string, params?: Record<string, unknown>): Promise<JsonRpcResponse> {
    this.#nextId += 1;
    const message: Record<string, unknown> = { jsonrpc: '2.0', id: this.#nextId, method };
    if (params !== undefined) message.params = params;
    this.scanner.add('mcp_traffic', JSON.stringify(message));
    this.scanner.add('mcp_tool_arguments', JSON.stringify(params ?? {}));

    const response = await this.server.handleMessage(message);
    const rendered = JSON.stringify(response);
    this.scanner.add('mcp_traffic', rendered);
    this.scanner.add('mcp_tool_results', rendered);
    if (!response) throw new Error(`no response for ${method}`);
    return response;
  }

  async callTool(name: string, args: Record<string, unknown>): Promise<Record<string, unknown>> {
    const response = await this.rpc('tools/call', { name, arguments: args });
    const result = response.result;
    if (!result) throw new Error(`tool call failed: ${JSON.stringify(response)}`);
    return result.structuredContent as Record<string, unknown>;
  }

  // -- HTTP -------------------------------------------------------------------

  async get(path: string): Promise<HttpResult> {
    this.scanner.add('http_urls', path);
    const response = await fetch(`${this.ui.baseUrl ?? ''}${path}`, {
      redirect: 'manual',
      headers: { host: '127.0.0.1' },
    });
    const body = await response.text();
    this.scanner.add('ui_html', body);
    return { status: response.status, headers: response.headers, body };
  }

  async post(path: string, fields: Record<string, string>): Promise<HttpResult> {
    this.scanner.add('http_urls', path);
    const body = new URLSearchParams(fields).toString();
    const response = await fetch(`${this.ui.baseUrl ?? ''}${path}`, {
      method: 'POST',
      redirect: 'manual',
      headers: { 'content-type': 'application/x-www-form-urlencoded', host: '127.0.0.1' },
      body,
    });
    const text = await response.text();
    if (response.status !== 303) this.scanner.add('ui_html', text);
    return { status: response.status, headers: response.headers, body: text };
  }

  // -- flows ------------------------------------------------------------------

  pathFor(payload: Record<string, unknown>): string {
    const url = payload.authorization_url;
    if (typeof url !== 'string') throw new Error('the payload carries no authorization_url');
    return new URL(url).pathname;
  }

  submitSecret(payload: Record<string, unknown>, value: string): Promise<HttpResult> {
    return this.post(`${this.pathFor(payload)}/submit`, { secret: value });
  }

  async confirm(payload: Record<string, unknown>): Promise<HttpResult> {
    const path = this.pathFor(payload);
    const page = await this.get(path);
    const token = /name="confirm_token" value="([^"]+)"/.exec(page.body)?.[1];
    if (!token) throw new Error('stage B page did not render a confirmation token');
    return this.post(`${path}/confirm`, { confirm_token: token });
  }

  cancel(payload: Record<string, unknown>): Promise<HttpResult> {
    return this.post(`${this.pathFor(payload)}/cancel`, {});
  }

  status(requestId: string): Promise<Record<string, unknown>> {
    return this.callTool('secret_status', { request_id: requestId });
  }

  /** Stage A (+ Stage B when required) through to a terminal state. */
  async finishFlow(
    payload: Record<string, unknown>,
    value: string,
  ): Promise<Record<string, unknown>> {
    await this.get(this.pathFor(payload));
    await this.submitSecret(payload, value);
    let status = await this.status(payload.request_id as string);
    if (status.state === 'AWAITING_EXECUTION_CONFIRMATION') {
      await this.confirm(payload);
      status = await this.status(payload.request_id as string);
    }
    return status;
  }

  collectLogs(): void {
    this.scanner.add('application_logs', this.logText);
    this.scanner.add('audit_records', JSON.stringify(this.logRecords));
  }
}

export function storeArgs(overrides: Record<string, unknown> = {}): Record<string, unknown> {
  return {
    destination: 'fake-store',
    name: 'STRIPE_SECRET_KEY',
    target: { project: 'acme-dev-project', secret: 'STRIPE_SECRET_KEY' },
    write_mode: 'create',
    environment: 'development',
    ...overrides,
  };
}
